flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/DynamicWriter.java

```diff
@@ -116,6 +117,11 @@ public void write(DynamicRecordInternal element, Context context)
                           Preconditions.checkState(
                               !equalityFieldIds.isEmpty(),
                               "Equality field columns shouldn't be empty when configuring to use UPSERT data.");
+
+                          Preconditions.checkArgument(
+                              !(TableUtil.formatVersion(table) > 2),
+                              "Dynamic Sink writer does not support upsert mode in tables (V3+)");
```

Comment:
Maybe put a little more information, e.g. "Dynamic Sink writer does not currently support upsert mode with deletion vectors in V3+ tables. Use append mode or configure the table to use position delete files."?